sql/validation.go

```
@@ -20,38 +20,82 @@ import (
-	tbl := randomTableName()
-	createStmt := fmt.Sprintf("CREATE TABLE %s LIFECYCLE %d AS SELECT *, RAND() AS %s FROM (%s) AS %s_ori", tbl, temporaryTableLifecycle, randomColumn, standardSelect, tbl)
-	if _, e := db.Exec(createStmt); e != nil {
-		log.Errorf("create temporary table failed, statememnt:[%s], err:%v", createStmt, e)
-		return "", e
+func releaseTrainingDataset(ds trainingDataset) {
+	// TODO(weiguo): release resources for databases, like: "hive", "mysql"...
+}
+
+func createMaxcomputeDataset(db *DB, slct string, trainingUpperbound float32) (trainingDataset, error) {
+	ds := namingTrainingDataset()
+	// create a table, then split it into 2 views
+	stmt := fmt.Sprintf("CREATE TABLE %s LIFECYCLE %d AS SELECT *, RAND() AS %s FROM (%s) AS %s_ori", ds.table, temporaryTableLifecycle, randomColumn, slct, ds.table)
+	if _, e := db.Exec(stmt); e != nil {
+		log.Errorf("create temporary table failed, stmt:[%s], err:%v", stmt, e)
+		return trainingDataset{}, e
 	}
-	return tbl, nil
+	trainingCond := fmt.Sprintf("%s < %f", randomColumn, trainingUpperbound)
+	if e := createView(db, ds.table, ds.trainingView, trainingCond); e != nil {
+		return trainingDataset{}, e
+	}
+	validationCond := fmt.Sprintf("%s >= %f", randomColumn, trainingUpperbound)
+	if e := createView(db, ds.table, ds.validationView, validationCond); e != nil {
+		return trainingDataset{}, e
+	}
+	return ds, nil
+}
+
+func createView(db *DB, table, view, where string) error {
+	stmt := fmt.Sprintf("CREATE VIEW %s AS SELECT * FROM %s WHERE %s", view, table, where)
+	if _, e := db.Exec(stmt); e != nil {
+		log.Errorf("create view failed, stmt:[%s], err:%v", stmt, e)
+		return e
+	}
+	return nil
 }
 
-func randomTableName() string {
-	n := time.Now().UnixNano() / 1e3
-	return fmt.Sprintf("%s%d", randomTablePrefix, n)
+func namingTrainingDataset() trainingDataset {
+	uniq := time.Now().UnixNano() / 1e3
+	return trainingDataset{
+		supported:      true,
```

Is `supported` field necessary?

- If a database DOES support `namingTrainingDataset`, `supported` field is always true.
- If a database DOES NOT support `namingTrainingDataset`, the function should raise an error.

---

`supported` is a temporary variable for `codegen`. Here it used to avoid damage to `CI`

https://github.com/weiguoz/sqlflow/blob/5618433e2490ed50b72d48af06d77eef6d18ca72/sql/validation.go#L24-L27